RecoEcal/EgammaCoreTools/src/EcalClustersGraph.cc

```diff
@@ -0,0 +1,423 @@
+      double dphi = deltaPhi(seed_phi, cl_phi);
+      double deta = deltaEta(seed_eta, cl_eta);
+
+      if (seedLocal[2] == clusterLocal[2] && deta >= width[0] && deta <= width[1] && std::abs(dphi) <= width[2]) {
+        graphMap_.addEdge(is, icl);
+      }
+    }
+  }
+}
+
+std::vector<std::vector<double>> EcalClustersGraph::fillHits(const CaloCluster* cluster) {
+  const std::vector<std::pair<DetId, float>>& hitsAndFractions = cluster->hitsAndFractions();
+  std::vector<std::vector<double>> out(hitsAndFractions.size());
+  if (hitsAndFractions.empty()) {
+    edm::LogError("EcalClustersGraph") << "No hits in cluster!!";
+  }
+  for (unsigned int i = 0; i < hitsAndFractions.size(); i++) {
+    std::vector<double> rechit(DeepSCConfiguration::nRechitsFeatures);
+    if (hitsAndFractions[i].first.subdetId() == EcalBarrel) {
+      double energy = (*recHitsEB_->find(hitsAndFractions[i].first)).energy();
+      EBDetId eb_id(hitsAndFractions[i].first);
+      rechit[0] = eb_id.ieta();                         //ieta
+      rechit[1] = eb_id.iphi();                         //iphi
+      rechit[2] = 0.;                                   //iz
+      rechit[3] = energy * hitsAndFractions[i].second;  //energy * fraction
+    } else if (hitsAndFractions[i].first.subdetId() == EcalEndcap) {
+      double energy = (*recHitsEE_->find(hitsAndFractions[i].first)).energy();
+      EEDetId ee_id(hitsAndFractions[i].first);
+      rechit[0] = ee_id.ix();  //ix
+      rechit[1] = ee_id.iy();  //iy
+      if (ee_id.zside() < 0)
+        rechit[2] = -1.;  //iz
+      if (ee_id.zside() > 0)
+        rechit[2] = +1.;                                //iz
+      rechit[3] = energy * hitsAndFractions[i].second;  //energy * fraction
+    } else {
+      edm::LogError("EcalClustersGraph") << "Rechit is not either EB or EE!!";
+    }
+    out[i] = rechit;
+  }
+  return out;
+}
+
+std::vector<double> EcalClustersGraph::computeVariables(const CaloCluster* seed, const CaloCluster* cluster) {
+  std::vector<double> cl_vars(12);  //TODO dynamic configuration
+  const auto& clusterLocal = clusterPosition(cluster);
+  cl_vars[0] = cluster->energy();                              //cl_energy
+  cl_vars[1] = cluster->energy() / std::cosh(cluster->eta());  //cl_et
+  cl_vars[2] = cluster->eta();                                 //cl_eta
+  cl_vars[3] = cluster->phi();                                 //cl_phi
+  cl_vars[4] = clusterLocal[0];                                //cl_ieta/ix
+  cl_vars[5] = clusterLocal[1];                                //cl_iphi/iy
+  cl_vars[6] = clusterLocal[2];                                //cl_iz
+  cl_vars[7] = deltaEta(seed->eta(), cluster->eta());          //cl_dEta
+  cl_vars[8] = deltaPhi(seed->phi(), cluster->phi());          //cl_dPhi
+  cl_vars[9] = seed->energy() - cluster->energy();             //cl_dEnergy
+  cl_vars[10] = (seed->energy() / std::cosh(seed->eta())) - (cluster->energy() / std::cosh(cluster->eta()));  //cl_dEt
```

Comment:
```suggestion
  cl_vars[10] = (seed->energy() / std::cosh(seed->eta())) - cl_vars[1]);  //cl_dEt
```
could avoid computung twice the (expensive)  `1 / std::cosh(cluster->eta())`

I have the impression, however, that the net should be able to identify the correlations by itself, and perhaps the same results could be obtained by putting in `cl_vars[7-10]` the quantities for the seed, without doing already the difference. (Not relevant. just taking note of it since I jumped here)

Reply:
I have removed the repeating accesses to the pointers. 

About the input variables: you're right that the model should learn the dependence on the difference, but there is no harm in helping it with the additional input.